### PR TITLE
enhancement(aws_sqs source): Optimize event send loop

### PR DESCRIPTION
**Pull request by bruceg**
This reduces the number of sends through the `SourceSender` to one per batch.

### CODE REVIEW

File: src/sources/aws_sqs/source.rs [hunk 1]
```diff
@@ -90,35 +90,33 @@ impl SqsSource {
         };
 
         if let Some(messages) = receive_message_output.messages {
-            let mut receipts_to_ack = vec![];
+            let mut receipts_to_ack = Vec::with_capacity(messages.len());
+            let mut events = Vec::new();
```
Comment (fuchsnj) [inline src/sources/aws_sqs/source.rs:94]: `events` should probably also be pre-allocated with `Vec::with_capacity(messages.len())`
While there could be more events, the most likely case is usually 1 event per message.

**Review by fuchsnj — APPROVED** [review]

**Review by pablosichert — APPROVED** [review]
Good catch! 👍

File: src/sources/aws_sqs/source.rs [hunk 2]
```diff
@@ -90,35 +90,33 @@ impl SqsSource {
...
-                            }
-                        }
+                        events.extend(decoded);
                     }
                 }
             }
             drop(batch);
+            emit!(&AwsSqsBytesReceived { byte_size });
+            if let Err(err) = out.send_batch(events).await {
+                error!(message = "Error sending to sink.", error = %err);
```
Comment (jszwedko) [inline src/sources/aws_sqs/source.rs:118]: Don't we have an internal event for this failed send case?

**Review by jszwedko — APPROVED** [review]
👍 One non-blocking comment.

File: src/sources/aws_sqs/source.rs (hunk 2 above)
Comment (bruceg) [reply src/sources/aws_sqs/source.rs:118]: Oops, yes, and I knew that. Good catch.

**Comment by bruceg** [comment]
I realized the source messages would end up getting deleted even when the channel was closed, so there's now a bug fix here too.

File: src/sources/aws_sqs/source.rs [hunk 3]
```diff
@@ -90,47 +90,47 @@ impl SqsSource {
...
-                            // a receipt handle should always exist
-                            if let Some(receipt_handle) = message.receipt_handle {
-                                receipts_to_ack.push(receipt_handle);
-                            }
-                        }
+                        events.extend(decoded);
                     }
                 }
             }
             drop(batch);
```
Comment (jszwedko) [inline src/sources/aws_sqs/source.rs:115]: This may not be relevant, but why do we drop the batch explicitly?

**Review by jszwedko — APPROVED** [review]
👍 good catch with the bug fix too.

File: src/sources/aws_sqs/source.rs (hunk 3 above)
Comment (bruceg) [reply src/sources/aws_sqs/source.rs:115]: To ensure there isn't an extraneous reference to the batch, which would deadlock acknowledgement handling.

**Comment by pablosichert** [comment]
@bruceg I'm wondering how this logic behaves when the stream closed, but the events array has already partially been sent. I suppose we don't acknowledge the batch and send the messages again (potentially multiple times)?

Not necessarily that we can do something about it, just curious. The previous code would probably do slightly better in that regard since it doesn't send one huge batch.

File: src/sources/aws_sqs/source.rs (hunk 3 above)
Comment (pablosichert) [reply src/sources/aws_sqs/source.rs:115]: @jszwedko will probably not the last one to stumble on that, I'd suggest to leave a comment for the next person now that we have the context 👍😄

**Comment by pablosichert** [comment]
Another question: When looking at that code on master, it also seems to have the bug you fixed in 29b5398acdfd76ec20d97c0a28fb3ad23a769ac4. Should we maybe pull that out into another PR so that we can cover it in the release notes as a `fix`?

**Comment by bruceg** [comment]
> Should we maybe pull that out into another PR so that we can cover it in the release notes as a `fix`?

Agreed.

**Review by pablosichert — APPROVED** [review]

**Comment by bruceg** [comment]
> @bruceg I'm wondering how this logic behaves when the stream closed, but the events array has already partially been sent. I suppose we don't acknowledge the batch and send the messages again (potentially multiple times)?

While the potential is there for this to cause duplicate messages, it's pretty unlikely. The following all have to be true:
1. The batch turned into more than one `EventArray`, which means there were either more than 1024 events or more than one type of event.
2. At least one element of the batch was sent _before_ the channel was closed.
3. Shutdown handling on the channel failed to stop this source before closing.

For it to trigger more than one repeat would pretty much require some nasty bug in the above.

**Pull request merged**